R2API.Rules/RuleCatalogExtras.cs

```diff
@@ -0,0 +1,342 @@
+        {
+            RuleCategoryDef category = RuleCatalog.GetCategoryDef(ruleCategoryDefIndex);
+            AddRuleToCategoryInternal(category, ruleCategoryDefIndex, ruleDef);
+        }
+
+        static void AddRuleInternal(RuleDef ruleDef, int ruleCategoryDefIndex)
+        {
+            if (!categoryToCustomRules.ContainsKey(ruleCategoryDefIndex))
+            {
+                categoryToCustomRules[ruleCategoryDefIndex] = new List<RuleDef>();
+            }
+            categoryToCustomRules[ruleCategoryDefIndex].Add(ruleDef);
+        }
+    }
+
+    /// <summary>
+    /// Finds a custom ruleDef by comparing the ruleDef's global name with <paramref name="ruleDefGlobalName"/>
+    /// </summary>
+    /// <param name="ruleDefGlobalName">The global name of the ruleDef to find</param>
+    /// <returns>The rule def, null if it wasnt found</returns>
+    public static RuleDef FindCustomRuleDef(string ruleDefGlobalName)
+    {
+        SetHooks();
+
+        if(RuleCatalog.availability.available)
+        {
+            return RuleCatalog.FindRuleDef(ruleDefGlobalName);
+        }
+
+        return FindRuleDefInternal(ruleDefGlobalName);
+
+        static RuleDef FindRuleDefInternal(string ruleDefGlobalName)
+        {
+            foreach (RuleDef ruleDef in categoryToCustomRules.Values.SelectMany(x => x))
+            {
+                if (string.Equals(ruleDefGlobalName, ruleDef.globalName, StringComparison.OrdinalIgnoreCase))
+                {
+                    return ruleDef;
+                }
+            }
+            return null;
+        }
+    }
+
+    /// <summary>
+    /// Finds a custom RuleChoiceDef by comparing the ruleChoiceDef's global name with <paramref name="ruleChoiceDefGlobalName"/>
+    /// </summary>
+    /// <param name="ruleChoiceDefGlobalName">The global name of the rule choice def to find</param>
+    /// <returns>The rule choice def, null if it wasnt found</returns>
+    public static RuleChoiceDef FindCustomChoiceDef(string ruleChoiceDefGlobalName)
+    {
+        SetHooks();
+
+        if(RuleCatalog.availability.available)
+        {
+            return RuleCatalog.FindChoiceDef(ruleChoiceDefGlobalName);
+        }
+
+        return FindChoiceDefInternal(ruleChoiceDefGlobalName);
+
+        static RuleChoiceDef FindChoiceDefInternal(string ruleChoiceDefGlobalName)
+        {
+            foreach (RuleChoiceDef choiceDef in categoryToCustomRules.Values.SelectMany(x => x).SelectMany(x => x.choices))
+            {
+                if (string.Equals(ruleChoiceDefGlobalName, choiceDef.globalName, StringComparison.OrdinalIgnoreCase))
+                {
+                    return choiceDef;
+                }
+            }
+            return null;
+        }
+    }
+
+    #region hooks
+    internal static void SetHooks()
+    {
+        if(_hooksEnabled)
+        {
+            return;
+        }
+
+        
+        RuleCatalog.availability.CallWhenAvailable(FinishRulebookSetup);
```

Comment:
no unsub?

Reply:
CallWhenAvailable, which is the public exposed method, calls the action immediatly if "Available" is true, otherwise it adds it to the private action

Even then, MakeAvailable() calls the actions and then sets the action to null, so unsubbing is not necesary

Reply:
Yeah though I just realized, while this is unlikely, that the module could get enabled, disabled, and then re-enabled all while the catalog isn't yet available, triggering a bug since we never unsub